internal/corerad/advertise.go

```diff
@@ -487,15 +452,190 @@ func (a *Advertiser) buildRA(ifi config.Interface) (*ndp.RouterAdvertisement, er
 	// about that later.
 	ra.Options = append(ra.Options, &ndp.LinkLayerAddress{
 		Direction: ndp.Source,
-		Addr:      a.ifi.HardwareAddr,
+		Addr:      a.mac,
 	})
 
 	return ra, nil
 }
 
+// init initializes the Advertiser in preparation for handling NDP traffic.
+func (a *Advertiser) init() (bool, error) {
+	// Verify the interface is available and ready for listening.
+	ifi, err := a.checkInterface(a.iface)
+	if err != nil {
+		return false, err
+	}
+
+	// Initialize the NDP listener.
+	c, ip, err := a.dial(ifi)
+	if err != nil {
+		return false, fmt.Errorf("failed to create NDP listener: %v", err)
+	}
+
+	// We can now initialize any plugins that rely on dynamic information
+	// about the network interface.
+	for _, p := range a.cfg.Plugins {
+		if err := p.Prepare(ifi); err != nil {
+			return false, fmt.Errorf("failed to prepare plugin %q: %v", p.Name(), err)
+		}
+	}
+
+	// If possible, disable IPv6 autoconfiguration on this interface so that
+	// our RAs don't configure more IP addresses on this interface.
+	autoPrev, err := a.getIPv6Autoconf(a.iface)
+	if err != nil {
+		return false, fmt.Errorf("failed to get IPv6 autoconfiguration state on %q: %v", a.iface, err)
+	}
+
+	if err := a.setIPv6Autoconf(a.iface, false); err != nil {
+		if errors.Is(err, os.ErrPermission) {
+			// Continue anyway but provide a hint.
+			a.logf("permission denied while disabling IPv6 autoconfiguration, continuing anyway (try setting CAP_NET_ADMIN)")
+			a.mm.ErrorsTotal.WithLabelValues(a.iface, "configuration").Inc()
+		} else {
+			return false, fmt.Errorf("failed to disable IPv6 autoconfiguration on %q: %v", a.iface, err)
+		}
+	}
+
+	// We only want to accept router solicitation messages.
+	var f ipv6.ICMPFilter
+	f.SetAll(true)
+	f.Accept(ipv6.ICMPTypeRouterSolicitation)
+
+	if err := c.SetICMPFilter(&f); err != nil {
+		return false, fmt.Errorf("failed to apply ICMPv6 filter: %v", err)
+	}
+
+	// Enable inspection of IPv6 control messages.
+	flags := ipv6.FlagHopLimit
+	if err := c.SetControlMessage(flags, true); err != nil {
+		return false, fmt.Errorf("failed to apply IPv6 control message flags: %v", err)
+	}
+
+	// We are now a router.
+	if err := c.JoinGroup(net.IPv6linklocalallrouters); err != nil {
+		return false, fmt.Errorf("failed to join IPv6 link-local all routers multicast group: %v", err)
+	}
+
+	a.c = c
+	a.mac = ifi.HardwareAddr
+
+	a.logf("initialized, advertising from %s", ip)
+
+	return autoPrev, nil
+}
+
+// reinit attempts repeated reinitialization of the Advertiser based on whether
+// the input error is considered recoverbale.
+func (a *Advertiser) reinit(ctx context.Context, err error) (bool, error) {
+	// Notify the beginning and end of reinit logic to anyone listening.
+	notify := func() {
+		select {
+		case a.reinitC <- struct{}{}:
+		default:
+		}
+	}
+
+	notify()
+	defer notify()
+
+	// Check for conditions which are recoverable.
+	funcs := []func(err error) bool{
```

Comment:
This feels like a bit of premature complexity. L542 through L564 is a little hard to follow, compared to equivalent straight-line code, e.g.:

```go
var serr *os.SyscallError
if !(errors.As(err, &serr) || errors.Is(err, errLinkNotReady)) {
  // Unrecoverable error
  return false, err
}
```

Reply:
Fair. I think the trouble is that I want to inspect the error codes for things like EINVAL as I assume not all os.SyscallErrors would be recoverable conditions. This is going to require a bit more digging I think.